### PR TITLE
More arities for tuple returning zip extension method

### LINKED ISSUE #43702: More arities for tuple returning zip extension method
## Background and Motivation

https://github.com/dotnet/runtime/issues/43687#issuecomment-713714293

## Proposed API

```diff
namespace System.Linq
{
    public static class Enumerable
    {
+       public static IEnumerable<(TFirst First, TSecond Second, TThird Third)> Zip<TFirst, TSecond, TThird>(this IEnumerable<TFirst> first, IEnumerable<TSecond> second, IEnumerable<TThird> third);
    }

    public static class Queryable
    {
+       public static IQueryable<(TFirst First, TSecond Second, TThird Third)> Zip<TFirst, TSecond, TThird>(this IQueryable<TFirst> source1, IEnumerable<TSecond> source2, IEnumerable<TThird> source3);
    }
}
```

## Usage Examples

```c#
foreach(var (a, b, c) in first.Zip(second, third))
{
   //...
}
```

## Alternative Designs

```c#
public static IEnumerator<(T1, T2, T3)> GetEnumerator<T1, T2, T3>(this (IEnumerable<T1> a, IEnumerable<T2> b, IEnumerable<T3> c) pair)
    => pair.a.Zip(pair.b, (a, b) => (a, b)).Zip(pair.c, (a, c) => (a.a, a.b, c)).GetEnumerator();
```
```c#
foreach(var (a, b, c) in (collectionA, collectionB, collectionC)) 
{
	// This block will execute only while both `a`, `b` and `c` continue to yield values

	// Code goes here
}
```

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Closes #43702

It grows exponentially for test exhaustiveness. I'm just focusing on the third parameter.

### CODE REVIEW

File: src/libraries/System.Linq/tests/ZipTests.cs [hunk 1]
```diff
@@ -601,5 +601,123 @@ public void Zip2_TupleNames()
...
+            IEnumerable<int> first = new[] { 1, 2 };
+            IEnumerable<int> second = new[] { 3, 4 };
+            IEnumerable<int> third = new[] { 5, 6, 7 };
+            IEnumerable<(int, int, int)> expected = new[] { (1, 3, 5), (2, 4, 6) };
+
+            Assert.Equal(expected, first.Zip(second, third));
+        }
+
+        [Fact]
+        public void Zip3_ThridManyMore()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Linq/tests/ZipTests.cs:680]:
```suggestion
        public void Zip3_ThirdManyMore()
```

**Review by eiriktsarpalis — APPROVED** [review]
Looks great, thank you for the contribution!

File: src/libraries/System.Linq/src/System/Linq/Zip.cs [hunk 2]
```diff
@@ -42,6 +42,36 @@ public static IEnumerable<(TFirst First, TSecond Second)> Zip<TFirst, TSecond>(t
...
+        /// <summary>
+        /// Produces a sequence of tuples with elements from the three specified sequences.
+        /// </summary>
+        /// <typeparam name="TFirst">The type of the elements of the first input sequence.</typeparam>
+        /// <typeparam name="TSecond">The type of the elements of the second input sequence.</typeparam>
+        /// <typeparam name="TThird">The type of the elements of the third input sequence.</typeparam>
+        /// <param name="first">The first sequence to merge.</param>
+        /// <param name="second">The second sequence to merge.</param>
+        /// <param name="third">The third sequence to merge.</param>
+        /// <returns>A sequence of tuples with elements taken from the first, second and third sequences, in that order.</returns>
```
Comment (stephentoub) [inline src/libraries/System.Linq/src/System/Linq/Zip.cs:54]: Nit: long live the Oxford comma :)
```suggestion
        /// <returns>A sequence of tuples with elements taken from the first, second, and third sequences, in that order.</returns>
```

File: src/libraries/System.Linq/tests/ZipTests.cs [hunk 3]
```diff
@@ -601,5 +601,123 @@ public void Zip2_TupleNames()
...
         }
+
+        [Fact]
+        public void Zip3_FirstIsNull()
+        {
+            IEnumerable<int> first = null;
+            IEnumerable<int> second = new[] { 4, 5, 6 };
+            IEnumerable<int> third = new[] { 7, 8, 9 };
+
+            Assert.Throws<ArgumentNullException>("first", () => first.Zip(second, third));
```
Comment (stephentoub) [inline src/libraries/System.Linq/tests/ZipTests.cs:612]:
```suggestion
            AssertExtensions.Throws<ArgumentNullException>("first", () => first.Zip(second, third));
```

File: src/libraries/System.Linq/tests/ZipTests.cs [hunk 4]
```diff
@@ -601,5 +601,123 @@ public void Zip2_TupleNames()
...
+        }
+
+        [Fact]
+        public void Zip3_SecondIsNull()
+        {
+            IEnumerable<int> first = new[] { 1, 2, 3 };
+            IEnumerable<int> second = null;
+            IEnumerable<int> third = new[] { 4, 5, 6 };
+
+            Assert.Throws<ArgumentNullException>("second", () => first.Zip(second, third));
```
Comment (stephentoub) [inline src/libraries/System.Linq/tests/ZipTests.cs:622]:
```suggestion
            AssertExtensions.Throws<ArgumentNullException>("second", () => first.Zip(second, third));
```

File: src/libraries/System.Linq/tests/ZipTests.cs [hunk 5]
```diff
@@ -601,5 +601,123 @@ public void Zip2_TupleNames()
...
+        }
+
+        [Fact]
+        public void Zip3_ThirdIsNull()
+        {
+            IEnumerable<int> first = new[] { 1, 2, 3 };
+            IEnumerable<int> second = new[] { 4, 5, 6 };
+            IEnumerable<int> third = null;
+
+            Assert.Throws<ArgumentNullException>("third", () => first.Zip(second, third));
```
Comment (stephentoub) [inline src/libraries/System.Linq/tests/ZipTests.cs:632]:
```suggestion
            AssertExtensions.Throws<ArgumentNullException>("third", () => first.Zip(second, third));
```

File: src/libraries/System.Linq.Queryable/src/System/Linq/CachedReflection.cs [hunk 6]
```diff
@@ -843,6 +843,13 @@ internal static class CachedReflectionInfo
              (s_Zip_TFirst_TSecond_TResult_3 = new Func<IQueryable<object>, IEnumerable<object>, Expression<Func<object, object, object>>, IQueryable<object>>(Queryable.Zip).GetMethodInfo().GetGenericMethodDefinition()))
               .MakeGenericMethod(TFirst, TSecond, TResult);
 
+        private static MethodInfo? s_Zip_TFirst_TSecond_TThird_3;
+
+        public static MethodInfo Zip_TFirst_TSecond_TThird_3(Type TFirst, Type TSecond, Type TThird) =>
+             (s_Zip_TFirst_TSecond_TThird_3 ??
+             (s_Zip_TFirst_TSecond_TThird_3 = new Func<IQueryable<object>, IEnumerable<object>, IEnumerable<object>, IQueryable<(object, object, object)>>(Queryable.Zip).GetMethodInfo().GetGenericMethodDefinition()))
+              .MakeGenericMethod(TFirst, TSecond, TThird);
```
Comment (stephentoub) [inline src/libraries/System.Linq.Queryable/src/System/Linq/CachedReflection.cs:851]: I was going to comment saying `??=` could be used to simplify this, but I see you're just following the pattern used in a bunch of other cases above.  We can clean it up later.

**Review by stephentoub — APPROVED** [review]

**Review by huoyaoyuan — COMMENTED** [review]
PR feedbacks

**Comment by eiriktsarpalis** [comment]
Thanks for the contribution!

**Comment by Logerfo** [comment]
Thank you! In which runtime version is this feature supposed to be available?

**Comment by eiriktsarpalis** [comment]
That would be released with .NET 6 next November.

**Pull request merged**